### DISCUSSION TITLE
Where is the right place to fetch data once

### DISCUSSION THREAD

**Discussion by karkovich** [vercel/next.js#13701] [Help]
I have an e-commerce ssr application based on next.js.
I want to fetch data on server before render the application, but i don't want to fetch this data every user routing.

So for example:
User visit `www.myapp.com/home` => server will fetch some data and return html to the client
User click on `Link` to `www.myapp.com/products/1` => server won't fetch data because it's routing.

However, if user refresh page or visit  `www.myapp.com/products/1` => server should fetch data because it the first time

**Comment by Gabsii** [comment]
If you want to fetch data on the initial page load you might want to use `getServerSideProps`. The page will then be pre-rendered on each page reload. 

More info can be found here: 
https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

**Reply by karkovich** [reply]
thanks for response. but this is exactly the problem.  `getServerSideProp` will execute every user navigation. i want to fetch data only once. i'll give you more details:

i have an dynamic store website -> so visiting `storeA.store.com` or `storeB.store.com`
going to one(the same) nextjs application.

it means on visiting `storeA.store.com`, i have to parse the url and render the store regard to the `storeA` configuration(colors, logo, title). but I dont want to fetch it before every navigation.
 I guess i could fetch this configuration only once when user visting first page. then, share this data across pages without fetch it every navigation.

is it possible?

**Reply by harborhoffer** [reply]
Couldn't you use `getInitialProps` inside a custom _app.js to only fetch initial data if on the server and then stuff that data into some global state that is available to all pages when client side routing?

**Comment by Vadorequest** [comment]
It depends if you're using SSR or SSG, but in both case what you want is either `getServerSideProps` (SSR) or `getStaticProps` (SSG).

Here is a demo about how both work:
https://nrn-v2-mst-aptd-gcms-lcz-sty-c1.now.sh/fr

Maybe https://github.com/UnlyEd/next-right-now could inspire you.

- SSG example: https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/src/pages/%5Blocale%5D/examples/native-features/example-with-ssg.tsx
- SSR example: https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/src/pages/%5Blocale%5D/examples/native-features/example-with-ssr.tsx

**Reply by karkovich** [reply]
thanks for response I add comment on @Gabsii response.
please take a look

**Reply by Vadorequest** [reply]
We do something very similar with NRN, which is called "[Multi Single Tenancy](https://unlyed.github.io/next-right-now/concepts/tenancy.html)".

Basically, we use the exact same source code for all our tenants, and we deploy them individually.

Demo tenant 1: https://nrn-v2-mst-aptd-gcms-lcz-sty-c1.now.sh
Demo tenant 2: https://nrn-v2-mst-aptd-gcms-lcz-sty-c2.now.sh

We use SSG for this purpose, not SSR. This way, we pre-build all pages statically when building the app, and there are no additional request when an end-user navigates a site.

Check out the "SSG with Fallback" option if you need to build some of those static page dynamically. And "SSG with revalidate" if you need to rebuild static pages after the initial build, dynamically.

**Reply by karkovich** [reply]
thanks for detailed response! amazing.
However, it looks quite overkill to pre-render all stores while it could be more than 1K.
and it is self registration system so i have to re-build every time?

**Reply by Vadorequest** [reply]
Yes, unless you use the fallback option, which may be what you want.

This way, you can pre-render only a few subset of pages (or none at all), and have them getting dynamically generated upon the first user request. From then, they'll be statically served.

**Reply by karkovich** [reply]
@Vadorequest I found that `  static async getInitialProps(ctx)` inside _document.js called once... for example ===> Visiting `app.com`==> _document.js exectute `getInitialProps` ==> navigate inside app to another app => _document.js won't execute `getInitialProps` ===> refresh page ===> wil execture `getInitialProps`

could it be an option to make request once there?

**Reply by Vadorequest** [reply]
What happens in _document isn't forwarded to pages, it's a very special part of the framework.

**Reply by karkovich** [reply]
yeah it's not a good idea!
first, thanks for your response and help!

I still can't achieve my goal. however, it could help to clear my issue.
In docs next.js says it would execute `getServerSideProps`

1)When you request this page directly
2)When you request this page on client-side page

How can I send a request only when user visit page directly and not every navigation?

in my use case I have to load some meta-data before render the website. but for client side navigation i don't need to fetch this data again and again.

how can i do that?
 
https://nextjs.org/docs/basic-features/data-fetching#only-runs-on-server-side

**Reply by Vadorequest** [reply]
If your concern is "overfetching", then you could use a caching strategy.

I've implemented one a few days ago at https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-at-lcz-sty/src/utils/caching/hybridCache.ts with TTL option, and works cross-pages by storing the cache on the filesystem.

**Reply by karkovich** [reply]
Sure it's possible to do cache strategy(redis, custom cache, etc..) but it sounds quite redundant.
in my use case I read the url for first visit `store1.myapp.com`. and following subdomain `store1` I fetch the proper theme. (it could be 100k of subdomains in cache).

For me, ask cache always sounds overkill. Why doesn't Next.js expose some hook to fetch data only for first visiting and not each navigation?

**Reply by Vadorequest** [reply]
https://github.com/vercel/next.js/discussions/13701#discussioncomment-24840 actually answers your question, just check if `req` exists in your `getInitialProps` and if so then perform your fetching.

This way, client side won't perform fetching. (but you might need to store results from `getInitialProps` in memory, I'm not sure about that)

It's not exactly what you want to do, but it's what's closest/easiest IMHO.

**Reply by karkovich** [reply]
@harborhoffer 
first `getInitialProps` doesn't recommended by docs [reference](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps). second I try to use `getInitialProps` but I aways get the `context.req` 

```javascript

MyApp.getInitialProps = async (appContext) => {
  // calls page's `getInitialProps` and fills `appProps.pageProps`
  const appProps = await App.getInitialProps(appContext)
  console.log('app', appContext.ctx.req)

  return { ...appProps }
}
```

**Reply by Vadorequest** [reply]
`req` cannot be defined during CSR. That's how SSR/CSR is meant to be detected.

**Reply by karkovich** [reply]
https://github.com/karkovich/nextjs-example
take a look :/

**Comment by sromexs** [comment]
you can just do this i found this and in the component you should check if `isRequest` is false (that means this is not first time) then you can fetch data from client side with `useEffect`.
i found this solution but `getInitialProps` not recomended by `nextjs`. if you find real solution share with us.
if you want, send me your social media id like telegram or whatsup to communicate with me this is me email `sadraromexs@gmail.com`. 

```javascript
MyApp.getInitialProps = async ({ req }) => {
  let isRequest = false;

  if (req) {
    const data = fetch("http://somethin");
    isRequest = true;
  }

  return { isRequest, data };
};
```

**Comment by srikanth912** [comment]
We are also using Next.JS for multi-tenant [eCommerce platform](https://www.agnicart.com/) and we do the following:

Inside _app.js getInitialProps, we fetch all common properties required by all pages for a site and store them in Redux Store (like Store Details, Menus like headers/footers. Theme Settings, Analytics Settings, User details, User Cart, Store custom CSS ... all are part of single API call so performance is not concern too) so we don't need to fetch them again for every page render (unless they change for example user logging out or adding/remove the product to/from the cart). 

So when a user visits the site, the first render happens on the server, and Redux Store initial state will have all common data for that store/user.

I hope this helps.

**Reply by karkovich** [reply]
thanks for response. However, I expect next.js to give us some intuitive way to regonize client-side/server-side navigation.

**Reply by sromexs** [reply]
Bro i can handle this with getInitialProps but in docs said thats deprecated
i want proper way with get getServerSideProps

**Reply by karkovich** [reply]
@srikanth912 may I ask you about multi-tenant eCommerce platform? how did you handle custom customer subdomains/domains architecture? 
1) does your nextjs application parse the request host/url and fetch the related meta-data(theme settings, etc...)?
2) How do you allow your clients to connect their own domain(with SSL)? what strategy are you using for?

Thanks a lot!

**Reply by srikanth912** [reply]
@karkovich 
1) Yes, we get the hostname from req and call API to fetch related meta data for that domain and populate in Redux Store in _app.js so that information is available to all pages/components (and fetched only once).
2) Today we are using Letsencrypt for SSL. Once customer add their custom domain in dashboard, our backend will initate a job to get letsecnrypt for their domain. (If they are not using custom domain, we use wild card SSL certificate from Letsencrypt for our domain). Each customer gets a sub-domain and they need to point CNAME to that sub-domain in their DNS for SSL to work properly.

**Reply by karkovich** [reply]
@srikanth912  thanks for your response!
About SSL for custom domain.
Once customer add their custom domain i guess your backend job create a certificate. 
How do you connect the certificate to the configuration? (In case of nginx I guess you have to add block for each domain and restart nginx)? or you are using another  HTTP server and reverse proxy?

thanks a lot!

**Reply by saisaran96** [reply]
@srikanth912 Inside _app.js getInitialProps i called the fetch to update the global state it works fine but the issue is that fetch is called each time we navigate . How did you handle this ?

**Reply by saisaran96** [reply]
@karkovich Have you able to solve the issue ?

**Reply by filippofilip95** [reply]
@saisaran96 I'm having same problem here. Have you resolved it somehow ?

**Reply by Innei** [reply]
> @srikanth912 Inside _app.js getInitialProps i called the fetch to update the global state it works fine but the issue is that fetch is called each time we navigate . How did you handle this ?

Hi, I solve like this. 

```ts
let initData: any = null
// cache 5 mins
setInterval(() => {
  initData = null
}, 1000 * 60 * 5)
// @ts-ignore
App.getInitialProps = async (props: AppContext) => {
  const appProps = await NextApp.getInitialProps(props)

  
  initData = initData || (await Rest('Aggregate').get<AggregateResp>())

  return { ...appProps, initData }
}

```

https://github.com/mx-space/kami/blob/5deae64485a513196c57c74fadda0c060cd47adb/pages/_app.tsx?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L313

**Reply by Ploppy3** [reply]
Using getInitialProps disables static optimization and page prefetching. I don't see any solution to this at this date.

**Comment by Innei** [comment]
I think next.js should provide a method or hook like `beforeStartClient` or `createLoadHelpers` in [rakkasjs](https://rakkasjs.org/guide/customization-hooks)

**Comment by stigkj** [comment]
@karkovich It sounds like you mean that each product page should be static, i.e. only rendered once at first visit? And I guess you cannot render the pages on build, i.e. use the standard static generation support in Next.js. Then you could look into the Incremental Static Regeneration support, specifically https://nextjs.org/docs/basic-features/data-fetching#fallback-blocking.

Here is an example of how this could look like:
```
export const getStaticProps = ({ params }) => {
  const param = <extract some param from params>
  const res = await fetch('<some api>?p=param')
  const json = await res.json()

  return {
    props: {
      data: json,
    },
  }
}

export const getStaticPaths = () => {
  return {
    paths: [],
    fallback: 'blocking',
  }
}

export default function Page(props) {
  return <render something with the props>
}
```

This would make only the first fetch of a page do the render. Subsequent fetches would use the already rendered page.
